src/DateDuration.ts

```diff
@@ -0,0 +1,97 @@
+import { parse } from "tinyduration";
+import { LocalDateUnit } from "./LocalDate";
+
+export type DateDurationUnit = LocalDateUnit;
+
+export class DateDuration {
+  public readonly value: ValidDateDurationObject;
+
+  constructor(readonly input: DateDurationObject) {
+    if (!validateDurationObject(input)) {
+      throw new Error("Provided duration object is not valid");
+    }
+    this.value = input;
+  }
+
+  static from(value: string | DateDuration | DateDurationObject): DateDuration {
+    if (value instanceof DateDuration) {
+      return value;
+    }
+
+    if (typeof value === "string") {
+      return DateDuration.fromString(value);
+    }
+
+    return DateDuration.fromObject(value);
+  }
+
+  static fromString(string: string): DateDuration {
+    const parsedDuration = parse(string);
+    return DateDuration.fromObject({
+      years: parsedDuration.years,
+      months: parsedDuration.months,
+      weeks: parsedDuration.weeks,
+      days: parsedDuration.days,
+    });
+  }
+
+  static fromObject({
+    years = 0,
+    months = 0,
+    weeks = 0,
+    days = 0,
+  }: DateDurationObject): DateDuration {
+    return new DateDuration({ years, months, weeks, days });
+  }
+
+  format(): string {
+    const yearText = formatUnit(this.value.years, "year", "years");
+    const monthText = formatUnit(this.value.months, "month", "months");
+    const weekText = formatUnit(this.value.weeks, "week", "weeks");
+    const dayText = formatUnit(this.value.days, "day", "days");
+    return `${yearText}${monthText}${weekText}${dayText}`.trim();
+  }
+
+  toString(): string {
+    return this.format();
+  }
+
+  toISOString(): string {
+    const { years, months, weeks, days } = this.value;
+    return `P${years}Y${months}M${weeks}W${days}DT0H0M0S`;
+  }
+}
+
+type ValidDateDurationObject = {
+  years: number;
+  months: number;
+  weeks: number;
+  days: number;
+};
+
+type DateDurationObject = {
+  years?: number;
+  months?: number;
+  weeks?: number;
+  days?: number;
+};
+
+function validateDurationObject(
+  object: DateDurationObject
+): object is ValidDateDurationObject {
+  for (const value of Object.values(object)) {
+    if (!Number.isInteger(value)) {
+      throw new TypeError("DateDuration only accepts integer duration values");
+    }
+    if (value < 0) {
+      throw new Error("Duration values cannot be less than 0");
+    }
+  }
+
+  return true;
+}
+
+function formatUnit(value: number, singular: string, plural: string): string {
+  const unit = value === 1 ? singular : plural;
+  return value ? `${value} ${unit} ` : "";
```

Comment:
NIT: was a bit confused on the logic here as ternary seems like an unexpected way to test for `value != 0` I would rather
```suggestion
  return value === 0  "" ? `${value} ${unit} `;
```

Reply:
https://github.com/ailohq/date/commit/79085222cdd1ce1291b04594d628c5539554813f